OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+* SimTK::Real(alias for double).                                               *
+*-----------------------------------------------------------------------------*/
+template<typename ET = SimTK::Real> // Element datatype.
+class OpenSim::DataTable_ : public OpenSim::AbstractDataTable {
+public:
+  using size_t                     = std::size_t;
+  using string                     = std::string;
+  using MetaDataType               = std::unordered_map<string, 
+                                                        SimTK::AbstractValue*>;
+  using ColLabelsType              = std::unordered_map<string, size_t>;
+  using ColLabelsIterType          = ColLabelsType::iterator;
+  using ColLabelsConstIterType     = ColLabelsType::const_iterator;
+  using ColLabelsIterPairType      = std::pair<ColLabelsIterType,
+                                               ColLabelsIterType>;
+  using ColLabelsConstIterPairType = std::pair<ColLabelsConstIterType,
+                                               ColLabelsConstIterType>;
+
+  /// Construct empty DataTable.
+  DataTable_() :
+    data{}, metadata{}, col_ind{} {}
+
+  /// Construct DataTable of given shape populated with given value val. 
+  /// Default value for val is NaN.
+  /// \param nrows Number of rows.
+  /// \param ncols Number of columns.
+  /// \param val   Value to initialize all the entries to.
+  DataTable_(size_t nrows,
+             size_t ncols,
+             const ET& val = ET{SimTK::NaN}) :
+    data{int(nrows), int(ncols), val}, metadata{}, col_ind{} {}
+
+  /// Construct DataTable using an iterator(satisfying requirement of an 
+  /// input_iterator) which produces one entry at a time. The entries of 
+  /// DataTable_ are copy initialized using the values produced by the 
+  /// iterator. For example, specifying ROWWISE for parameter dir and 10 for 
+  /// parameter ndir will populate the DataTable one row at a time with each 
+  /// row's length taken to be 10 elements.
+  ///
+  /// \param first Beginning of range covered by the iterator.
+  /// \param last End of the range covered by the iterator.
+  /// \param ndir Extent of the dimention specified by parameter dir. 
+  /// \param dir Dimension to populate the DataTable. Populate row-wise or 
+  ///              column wise. See OpenSim::InputItDir for possible values.
+  /// \param allow_missing Allow for missing values. When set to false, this
+  ///                      function will throw if the input iterator fills up 
+  ///                      the last row/col partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws ZeroElements When input-iterator does not produce any elements.
+  ///                      That is first == last.
+  template<typename InputIt>
+  DataTable_(InputIt first,
+             typename std::enable_if<!std::is_integral<InputIt>::value,
+                                     InputIt>::type last,
```

Comment:
> I am using SFINAE to discard the iterator overload 

Thanks, I see. I'm worried about the effect on doxygen. You should generate it and take a look. If it is bad, it may be possible to work around it by using more overloads to avoid default parameters, so that the non-templatized signatures match exactly. In that case the templatized version won't be used. These header files are serving multiple purposes; generally the view from doxygen (what users will see) is the most critical since our users are generally much less accomplished C++ programmers than we are. Also these same methods get mangled by SWIG into Python and Matlab interfaces using the same documentation so we want to make sure there is some reasonably straightforward translation. However, that might not apply here if this is definitely a C++ only method. In that case we'll need to clarify that in the documentation somehow. We can discuss further later and figure out what to do.